tests/unit/test_index_contents_parse.py

```diff
@@ -0,0 +1,426 @@
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_title=title_2, reference_value=value_2, rank=1
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=0, reference_title=title_3, reference_value=value_3, rank=2
+                ),
+            ),
+            id="many items second nested",
+        ),
+        pytest.param(
+            f"""# Contents
+- [{(title_1 := 'title 1')}]({(value_1 := 'value 1')})
+- [{(title_2 := 'title 2')}]({(value_2 := 'value 2')})
+  - [{(title_3 := 'title 3')}]({(value_3 := 'value 3')})
+""",
+            (
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=0, reference_title=title_1, reference_value=value_1, rank=0
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=0, reference_title=title_2, reference_value=value_2, rank=1
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_title=title_3, reference_value=value_3, rank=2
+                ),
+            ),
+            id="many items last nested",
+        ),
+        pytest.param(
+            f"""# Contents
+- [{(title_1 := 'title 1')}]({(value_1 := 'value 1')})
+  - [{(title_2 := 'title 2')}]({(value_2 := 'value 2')})
+  - [{(title_3 := 'title 3')}]({(value_3 := 'value 3')})
+""",
+            (
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=0, reference_title=title_1, reference_value=value_1, rank=0
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_title=title_2, reference_value=value_2, rank=1
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_title=title_3, reference_value=value_3, rank=2
+                ),
+            ),
+            id="many items nested",
+        ),
+        pytest.param(
+            f"""# Contents
+- [{(title_1 := 'title 1')}]({(value_1 := 'value 1')})
+  - [{(title_2 := 'title 2')}]({(value_2 := 'value 2')})
+    - [{(title_3 := 'title 3')}]({(value_3 := 'value 3')})
+""",
+            (
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=0, reference_title=title_1, reference_value=value_1, rank=0
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_title=title_2, reference_value=value_2, rank=1
+                ),
+                factories.IndexParsedListItemFactory(
+                    whitespace_count=4, reference_title=title_3, reference_value=value_3, rank=2
+                ),
+            ),
+            id="many items deeply nested",
```

Comment:
We should prehaps try deeply nested with different leaders

Reply:
good idea, should be fixed on next push